{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/coywil26/DIMPLE/blob/master/DIMPLE.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "FHy8rovTWDVB"
      },
      "outputs": [],
      "source": [
        "#@title 1. Install\n",
        "\n",
        "#@markdown Please execute each cell by pressing the *Play* button on the left.\n",
        "\n",
        "# download github \"https://github.com/coywil26/DIMPLE\" and install dependencies\n",
        "!git clone https://github.com/coywil26/DIMPLE\n",
        "%cd DIMPLE\n",
        "!pip install -r requirements.txt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "dbA01B3wWfMP"
      },
      "outputs": [],
      "source": [
        "#@title 2. Upload target file\n",
        "#@markdown Please execute to upload file\n",
        "from google.colab import files\n",
        "\n",
        "uploaded = files.upload()\n",
        "\n",
        "for fn in uploaded.keys():\n",
        "  print('User uploaded file \"{name}\" with length {length} bytes'.format(\n",
        "      name=fn, length=len(uploaded[fn])))\n",
        "target_file = fn\n",
        "!mkdir workspace\n",
        "!mv $target_file workspace\n",
        "directory = \"workspace/\""
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "J-ZYosDxWqZR",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "#@title 3. Generate Mutation Fragments and Primers\n",
        "#@markdown Oligo Parameters\n",
        "from google.colab import files\n",
        "import os\n",
        "import argparse\n",
        "from DIMPLE.DIMPLE import align_genevariation, print_all, post_qc, addgene, DIMPLE, generate_DMS_fragments\n",
        "from DIMPLE.utilities import parse_custom_mutations, codon_usage\n",
        "from Bio.Seq import Seq\n",
        "from google.colab import data_table\n",
        "import ast\n",
        "import pandas as pd\n",
        "\n",
        "\n",
        "# if target_file doesnt exist use default test file\n",
        "if 'target_file' not in locals():\n",
        "  target_file = \"DIMPLE/tests/combined_fasta.fa\"\n",
        "  wDir = \"DIMPLE/tests/\"\n",
        "  print('test run')\n",
        "else:\n",
        "  #wDir = os.path.basename(target_file)\n",
        "  wDir = directory\n",
        "\n",
        "geneFile = target_file\n",
        "\n",
        "oligoLen = 250 #@param {type:\"integer\"}\n",
        "fragmentLen = \"auto\"  #@param {type:\"string\"}\n",
        "if fragmentLen == \"auto\":\n",
        "  fragmentLen = \"\"\n",
        "melting_temp_low = 58 #@param {type:\"integer\"}\n",
        "melting_temp_high = 62 #@param {type:\"integer\"}\n",
        "overlap = 4 #@param {type:\"integer\"}\n",
        "barcode_start = 0 #@param {type:\"integer\"}\n",
        "#@markdown Use the following format for restriction enzymes<br>\n",
        "#@markdown BsmBI: CGTCTC(G)1/4<br>\n",
        "#@markdown BsaI: GGTCTC(G)1/4\n",
        "restriction_sequence = \"CGTCTC(G)1/4\" #@param {type:\"string\"}\n",
        "sequences_to_avoid = \"CGTCTC, GGTCTC\" #@param {type:\"string\"}\n",
        "avoid_sequence = sequences_to_avoid.split(\", \")\n",
        "matchSequences = False\n",
        "\n",
        "#@markdown Codon Usage (If using custom usage you will see a button to upload file at the bottom)\n",
        "usage = \"human\" #@param [\"human\", \"ecoli\"]\n",
        "custom_codon_usage = False #@param {type:\"boolean\"}\n",
        "# input custom mutation file\n",
        "if custom_codon_usage:\n",
        "  print(\"Look in DIMPLE/data/custom_codon_usage.txt for example file\")\n",
        "  uploaded = files.upload()\n",
        "  usage_file = list(uploaded.keys())[0]\n",
        "  with open(usage_file) as f:\n",
        "    usage = f.readlines()\n",
        "  usage = ast.literal_eval(usage[0].strip('\\n'))\n",
        "  pd.DataFrame(usage,index=[0])\n",
        "\n",
        "#@markdown Select Mutations (Select one or more)\n",
        "\n",
        "select_domain_insertion = False #@param {type:\"boolean\"}\n",
        "if select_domain_insertion:\n",
        "  dis = True\n",
        "else:\n",
        "  dis = False\n",
        "handle = \"AGCGGGAGACCGGGGTCTCTGAGC\" #@param {type:\"string\"}\n",
        "select_insertions = False #@param {type:\"boolean\"}\n",
        "insertions = \"GGG,GGGGGG\" #@param {type:\"string\"}\n",
        "if select_insertions:\n",
        "  insertions = insertions.split(\",\")\n",
        "else:\n",
        "  insertions = \"\"\n",
        "select_deletions = False #@param {type:\"boolean\"}\n",
        "deletions = \"3,6\" #@param {type:\"string\"}\n",
        "if select_deletions:\n",
        "  deletions = deletions.split(\",\")\n",
        "else:\n",
        "  deletions = \"\"\n",
        "select_deep_mutation_scan = True #@param {type:\"boolean\"}\n",
        "if select_deep_mutation_scan:\n",
        "  dms = True\n",
        "else:\n",
        "  dms = False\n",
        "#@markdown Mutation Settings\n",
        "maximize_nucleotide_change = True #@param {type:\"boolean\"}\n",
        "amino_acid_substitutions = \"Cys,Asp,Ser,Gln,Met,Asn,Pro,Lys,Thr,Phe,Ala,Gly,Ile,Leu,His,Arg,Trp,Val,Glu,Tyr\" #@param {type:\"string\"}\n",
        "include_stop_codons = False #@param {type:\"boolean\"}\n",
        "include_synonymous_mutations = False #@param {type:\"boolean\"}\n",
        "\n",
        "make_double_mutations = False #@param {type:\"boolean\"}\n",
        "select_custom_mutations = False #@param {type:\"boolean\"}\n",
        "if select_custom_mutations:\n",
        "  print(\"Look in DIMPLE/data/custom_codon_usage.txt for example file\")\n",
        "  uploaded = files.upload()\n",
        "  custom_mutations_file = list(uploaded.keys())[0]\n",
        "  with open(custom_mutations_file) as f:\n",
        "    lines = f.readlines()\n",
        "  custom_mutations = parse_custom_mutations(lines[1:])\n",
        "  pd.DataFrame(custom_mutations,index=[0])\n",
        "else:\n",
        "  custom_mutations = {}\n",
        "avoid_breaksites = False #@param {type:\"boolean\"}\n",
        "avoid_breaksites_list = \"\" #@param {type:\"string\"}\n",
        "\n",
        "#!python3 DIMPLE/run_dimple.py -geneFile=$target_file -oligoLen=$oligo_length -DMS=$select_DMS -usage=$codon_usage_bias -include_stop_codons=$include_stop_codons -restriction_sequence=$type_IIS_restriction_enzyme -avoid_sequence=$sequences_to_avoid -insertions=$insertions -deletions=$deletions -fragmentLen=$fragment_length -overlap=$fragment_overlap -barcode_start=$barcode_start_position -include_stop_codons=$include_stop_codons -usage=$custom_mutations\n",
        "\n",
        "\n",
        "# Check genetic handle for non nucleic characters\n",
        "if any([x not in ['A', 'C', 'G', 'T', 'a', 'c', 'g', 't'] for x in handle]):\n",
        "    raise ValueError('Genetic handle contains non nucleic bases')\n",
        "\n",
        "DIMPLE.handle = handle\n",
        "DIMPLE.synth_len = oligoLen\n",
        "if fragmentLen:\n",
        "    DIMPLE.maxfrag = fragmentLen\n",
        "else:\n",
        "    DIMPLE.maxfrag = oligoLen - 62 - overlap  # 62 allows for cutsites and barcodes\n",
        "\n",
        "DIMPLE.dms = dms\n",
        "\n",
        "#  adjust primer primerBuffer\n",
        "DIMPLE.primerBuffer += overlap\n",
        "\n",
        "DIMPLE.avoid_sequence = avoid_sequence\n",
        "DIMPLE.barcodeF = DIMPLE.barcodeF[int(barcode_start):]\n",
        "DIMPLE.barcodeR = DIMPLE.barcodeR[int(barcode_start):]\n",
        "tmp_cutsite = restriction_sequence.split('(')\n",
        "DIMPLE.cutsite = Seq(tmp_cutsite[0])\n",
        "DIMPLE.cutsite_buffer = Seq(tmp_cutsite[1].split(')')[0])\n",
        "tmp_overhang = tmp_cutsite[1].split(')')[1].split('/')\n",
        "DIMPLE.cutsite_overhang = int(tmp_overhang[1]) - int(tmp_overhang[0])\n",
        "DIMPLE.avoid_sequence = [Seq(x) for x in avoid_sequence]\n",
        "DIMPLE.stop_codon = include_stop_codons\n",
        "DIMPLE.make_double = make_double_mutations\n",
        "DIMPLE.usage = codon_usage(usage)\n",
        "DIMPLE.gene_primerTM = (melting_temp_low, melting_temp_high)\n",
        "DIMPLE.maximize_nucleotide_change = maximize_nucleotide_change\n",
        "\n",
        "OLS = addgene(os.path.join(wDir, geneFile).strip())\n",
        "if avoid_breaksites_list and select_custom_mutations:\n",
        "    OLS[0].problemsites = set(int(x) for x in custom_mutations.keys())\n",
        "    # add extras\n",
        "    if avoid_breaksites_list != '':\n",
        "        OLS[0].problemsites.update([int(x) for x in avoid_breaksites_list.split(',')])\n",
        "    for i in range(len(OLS[0].breaksites)):\n",
        "        DIMPLE.switch_fragmentsize(OLS[0], 1, OLS)\n",
        "if matchSequences == 'match':\n",
        "    align_genevariation(OLS)\n",
        "if deletions:\n",
        "    deletions = [int(x) for x in deletions]\n",
        "if not any([dis, dms, insertions, deletions]):\n",
        "    raise ValueError(\"Didn't select any mutations to generate\")\n",
        "print(dms)\n",
        "print(overlap)\n",
        "generate_DMS_fragments(OLS, overlap, overlap, include_synonymous_mutations, custom_mutations, dms, insertions, deletions, dis, wDir)\n",
        "\n",
        "post_qc(OLS)\n",
        "print_all(OLS, wDir)\n",
        "\n",
        "directory = wDir\n",
        "!zip -r dimple_results.zip $directory\n",
        "files.download(\"dimple_results.zip\")"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.10"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
