{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/coywil26/DIMPLE/blob/master/DIMPLE.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "cellView": "form",
        "id": "FHy8rovTWDVB",
        "outputId": "4374b03c-f6df-4244-e28a-8aa751d869ea",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Cloning into 'DIMPLE'...\n",
            "remote: Enumerating objects: 575, done.\u001b[K\n",
            "remote: Counting objects: 100% (220/220), done.\u001b[K\n",
            "remote: Compressing objects: 100% (142/142), done.\u001b[K\n",
            "remote: Total 575 (delta 152), reused 115 (delta 78), pack-reused 355\u001b[K\n",
            "Receiving objects: 100% (575/575), 1.23 MiB | 3.14 MiB/s, done.\n",
            "Resolving deltas: 100% (347/347), done.\n",
            "/content/DIMPLE\n",
            "Collecting biopython (from -r requirements.txt (line 1))\n",
            "  Downloading biopython-1.83-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m9.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 2)) (1.25.2)\n",
            "Installing collected packages: biopython\n",
            "Successfully installed biopython-1.83\n"
          ]
        }
      ],
      "source": [
        "#@title 1. Install\n",
        "\n",
        "#@markdown Please execute each cell by pressing the *Play* button on the left.\n",
        "\n",
        "# download github \"https://github.com/coywil26/DIMPLE\" and install dependencies\n",
        "!git clone https://github.com/coywil26/DIMPLE --branch maximize-nucleotide-mutations\n",
        "%cd DIMPLE\n",
        "!pip install -r requirements.txt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "cellView": "form",
        "id": "dbA01B3wWfMP",
        "outputId": "063efea6-5ead-45bf-9066-4d82bc660a58",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 90
        }
      },
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "     <input type=\"file\" id=\"files-d854aca5-4b7a-405b-84cd-d03b867e8576\" name=\"files[]\" multiple disabled\n",
              "        style=\"border:none\" />\n",
              "     <output id=\"result-d854aca5-4b7a-405b-84cd-d03b867e8576\">\n",
              "      Upload widget is only available when the cell has been executed in the\n",
              "      current browser session. Please rerun this cell to enable.\n",
              "      </output>\n",
              "      <script>// Copyright 2017 Google LLC\n",
              "//\n",
              "// Licensed under the Apache License, Version 2.0 (the \"License\");\n",
              "// you may not use this file except in compliance with the License.\n",
              "// You may obtain a copy of the License at\n",
              "//\n",
              "//      http://www.apache.org/licenses/LICENSE-2.0\n",
              "//\n",
              "// Unless required by applicable law or agreed to in writing, software\n",
              "// distributed under the License is distributed on an \"AS IS\" BASIS,\n",
              "// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n",
              "// See the License for the specific language governing permissions and\n",
              "// limitations under the License.\n",
              "\n",
              "/**\n",
              " * @fileoverview Helpers for google.colab Python module.\n",
              " */\n",
              "(function(scope) {\n",
              "function span(text, styleAttributes = {}) {\n",
              "  const element = document.createElement('span');\n",
              "  element.textContent = text;\n",
              "  for (const key of Object.keys(styleAttributes)) {\n",
              "    element.style[key] = styleAttributes[key];\n",
              "  }\n",
              "  return element;\n",
              "}\n",
              "\n",
              "// Max number of bytes which will be uploaded at a time.\n",
              "const MAX_PAYLOAD_SIZE = 100 * 1024;\n",
              "\n",
              "function _uploadFiles(inputId, outputId) {\n",
              "  const steps = uploadFilesStep(inputId, outputId);\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  // Cache steps on the outputElement to make it available for the next call\n",
              "  // to uploadFilesContinue from Python.\n",
              "  outputElement.steps = steps;\n",
              "\n",
              "  return _uploadFilesContinue(outputId);\n",
              "}\n",
              "\n",
              "// This is roughly an async generator (not supported in the browser yet),\n",
              "// where there are multiple asynchronous steps and the Python side is going\n",
              "// to poll for completion of each step.\n",
              "// This uses a Promise to block the python side on completion of each step,\n",
              "// then passes the result of the previous step as the input to the next step.\n",
              "function _uploadFilesContinue(outputId) {\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  const steps = outputElement.steps;\n",
              "\n",
              "  const next = steps.next(outputElement.lastPromiseValue);\n",
              "  return Promise.resolve(next.value.promise).then((value) => {\n",
              "    // Cache the last promise value to make it available to the next\n",
              "    // step of the generator.\n",
              "    outputElement.lastPromiseValue = value;\n",
              "    return next.value.response;\n",
              "  });\n",
              "}\n",
              "\n",
              "/**\n",
              " * Generator function which is called between each async step of the upload\n",
              " * process.\n",
              " * @param {string} inputId Element ID of the input file picker element.\n",
              " * @param {string} outputId Element ID of the output display.\n",
              " * @return {!Iterable<!Object>} Iterable of next steps.\n",
              " */\n",
              "function* uploadFilesStep(inputId, outputId) {\n",
              "  const inputElement = document.getElementById(inputId);\n",
              "  inputElement.disabled = false;\n",
              "\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  outputElement.innerHTML = '';\n",
              "\n",
              "  const pickedPromise = new Promise((resolve) => {\n",
              "    inputElement.addEventListener('change', (e) => {\n",
              "      resolve(e.target.files);\n",
              "    });\n",
              "  });\n",
              "\n",
              "  const cancel = document.createElement('button');\n",
              "  inputElement.parentElement.appendChild(cancel);\n",
              "  cancel.textContent = 'Cancel upload';\n",
              "  const cancelPromise = new Promise((resolve) => {\n",
              "    cancel.onclick = () => {\n",
              "      resolve(null);\n",
              "    };\n",
              "  });\n",
              "\n",
              "  // Wait for the user to pick the files.\n",
              "  const files = yield {\n",
              "    promise: Promise.race([pickedPromise, cancelPromise]),\n",
              "    response: {\n",
              "      action: 'starting',\n",
              "    }\n",
              "  };\n",
              "\n",
              "  cancel.remove();\n",
              "\n",
              "  // Disable the input element since further picks are not allowed.\n",
              "  inputElement.disabled = true;\n",
              "\n",
              "  if (!files) {\n",
              "    return {\n",
              "      response: {\n",
              "        action: 'complete',\n",
              "      }\n",
              "    };\n",
              "  }\n",
              "\n",
              "  for (const file of files) {\n",
              "    const li = document.createElement('li');\n",
              "    li.append(span(file.name, {fontWeight: 'bold'}));\n",
              "    li.append(span(\n",
              "        `(${file.type || 'n/a'}) - ${file.size} bytes, ` +\n",
              "        `last modified: ${\n",
              "            file.lastModifiedDate ? file.lastModifiedDate.toLocaleDateString() :\n",
              "                                    'n/a'} - `));\n",
              "    const percent = span('0% done');\n",
              "    li.appendChild(percent);\n",
              "\n",
              "    outputElement.appendChild(li);\n",
              "\n",
              "    const fileDataPromise = new Promise((resolve) => {\n",
              "      const reader = new FileReader();\n",
              "      reader.onload = (e) => {\n",
              "        resolve(e.target.result);\n",
              "      };\n",
              "      reader.readAsArrayBuffer(file);\n",
              "    });\n",
              "    // Wait for the data to be ready.\n",
              "    let fileData = yield {\n",
              "      promise: fileDataPromise,\n",
              "      response: {\n",
              "        action: 'continue',\n",
              "      }\n",
              "    };\n",
              "\n",
              "    // Use a chunked sending to avoid message size limits. See b/62115660.\n",
              "    let position = 0;\n",
              "    do {\n",
              "      const length = Math.min(fileData.byteLength - position, MAX_PAYLOAD_SIZE);\n",
              "      const chunk = new Uint8Array(fileData, position, length);\n",
              "      position += length;\n",
              "\n",
              "      const base64 = btoa(String.fromCharCode.apply(null, chunk));\n",
              "      yield {\n",
              "        response: {\n",
              "          action: 'append',\n",
              "          file: file.name,\n",
              "          data: base64,\n",
              "        },\n",
              "      };\n",
              "\n",
              "      let percentDone = fileData.byteLength === 0 ?\n",
              "          100 :\n",
              "          Math.round((position / fileData.byteLength) * 100);\n",
              "      percent.textContent = `${percentDone}% done`;\n",
              "\n",
              "    } while (position < fileData.byteLength);\n",
              "  }\n",
              "\n",
              "  // All done.\n",
              "  yield {\n",
              "    response: {\n",
              "      action: 'complete',\n",
              "    }\n",
              "  };\n",
              "}\n",
              "\n",
              "scope.google = scope.google || {};\n",
              "scope.google.colab = scope.google.colab || {};\n",
              "scope.google.colab._files = {\n",
              "  _uploadFiles,\n",
              "  _uploadFilesContinue,\n",
              "};\n",
              "})(self);\n",
              "</script> "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Saving ptwist_EF1a_IL7_FLAG_PDGFR.fa to ptwist_EF1a_IL7_FLAG_PDGFR.fa\n",
            "User uploaded file \"ptwist_EF1a_IL7_FLAG_PDGFR.fa\" with length 5772 bytes\n"
          ]
        }
      ],
      "source": [
        "#@title 2. Upload target file\n",
        "#@markdown Please execute to upload file\n",
        "from google.colab import files\n",
        "\n",
        "uploaded = files.upload()\n",
        "\n",
        "for fn in uploaded.keys():\n",
        "  print('User uploaded file \"{name}\" with length {length} bytes'.format(\n",
        "      name=fn, length=len(uploaded[fn])))\n",
        "target_file = fn\n",
        "!mkdir workspace\n",
        "!mv $target_file workspace\n",
        "directory = \"workspace/\""
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "id": "J-ZYosDxWqZR",
        "cellView": "form",
        "outputId": "d7e59480-fb80-46b9-9295-a7445efd4ad4",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 728
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.10/dist-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Initial Fragment Sizes for:ptwist_EF1a_IL7_FLAG_PDGFR\n",
            "[150, 150, 150]\n",
            "False\n",
            "4\n",
            "New max fragment:154\n",
            "Non specific Fragment:1\n",
            "[147, 153, 150]\n",
            "[[37, 181], [184, 334], [337, 484]]\n",
            "--------------------------------- Analyzing Gene:ptwist_EF1a_IL7_FLAG_PDGFR ---------------------------------\n",
            "Creating Fragment:ptwist_EF1a_IL7_FLAG_PDGFR --- Fragment #1 AA:2-49\n",
            "found primers:  ATACGTCTCGCAGTCGGCCACTGCG 63.8 15\n",
            "found primers:  ATACGTCTCGGGATGTTTTTGTTCCGCGCG 62.2 10\n",
            "Found non-specific match at 11bp:\n",
            "match: GTGGGAGCCGCAGTGGCCGACTGCG\n",
            "primer:ATACGTCTCGCAGTCGGCCACTGCG Tm:31.4\n",
            "Barcodes used:128\n",
            "Barcodes Remaining:2872\n",
            "Creating Fragment:ptwist_EF1a_IL7_FLAG_PDGFR --- Fragment #2 AA:51-100\n",
            "found primers:  ATACGTCTCTTTATTCGCGTCGCAGATATGTCTC 64.4 6\n",
            "found primers:  ATACGTCTCTTAGGAGAAGCACAGCCCACAAAG 61.6 7\n",
            "Barcodes used:26\n",
            "Barcodes Remaining:2846\n",
            "Creating Fragment:ptwist_EF1a_IL7_FLAG_PDGFR --- Fragment #3 AA:102-150\n",
            "found primers:  ATACGTCTCGGCGGGCTTTCTCCCC 60.9 15\n",
            "found primers:  ATACGTCTCGGAGGCAGCGGTGGAGGC 63.3 13\n",
            "Barcodes used:234\n",
            "Barcodes Remaining:2612\n",
            "Running QC for barcode primer specificity\n",
            "Checking primer set:ptwist_EF1a_IL7_FLAG_PDGFR_oligoP_DMS-1\n",
            "Checking primer set:ptwist_EF1a_IL7_FLAG_PDGFR_oligoP_DMS-2\n",
            "Checking primer set:ptwist_EF1a_IL7_FLAG_PDGFR_oligoP_DMS-3\n",
            "No non-specific primers detected\n",
            "  adding: workspace/ (stored 0%)\n",
            "  adding: workspace/All_Oligos.fasta (deflated 95%)\n",
            "  adding: workspace/ptwist_EF1a_IL7_FLAG_PDGFR_DMS_Oligo_Primers.fasta (deflated 61%)\n",
            "  adding: workspace/All_Primers.fasta (deflated 69%)\n",
            "  adding: workspace/ptwist_EF1a_IL7_FLAG_PDGFR.fa (deflated 66%)\n",
            "  adding: workspace/ptwist_EF1a_IL7_FLAG_PDGFR_DMS_Gene_Primers.fasta (deflated 64%)\n",
            "  adding: workspace/ptwist_EF1a_IL7_FLAG_PDGFR_DMS_Oligos.fasta (deflated 95%)\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "download(\"download_426e5c78-2ee2-40b5-8ebf-4bdc6da128b1\", \"file.zip\", 8233)"
            ]
          },
          "metadata": {}
        }
      ],
      "source": [
        "#@title 3. Generate Mutation Fragments and Primers\n",
        "#@markdown Oligo Parameters\n",
        "from google.colab import files\n",
        "import os\n",
        "import argparse\n",
        "from DIMPLE.DIMPLE import align_genevariation, print_all, post_qc, addgene, DIMPLE, generate_DMS_fragments\n",
        "from DIMPLE.utilities import parse_custom_mutations, codon_usage\n",
        "from Bio.Seq import Seq\n",
        "from google.colab import data_table\n",
        "import ast\n",
        "import pandas as pd\n",
        "\n",
        "class arguments:\n",
        "  pass\n",
        "args = arguments()\n",
        "\n",
        "# if target_file doesnt exist use default test file\n",
        "if 'target_file' not in locals():\n",
        "  target_file = \"DIMPLE/tests/combined_fasta.fa\"\n",
        "  args.wDir = \"DIMPLE/tests/\"\n",
        "  print('test run')\n",
        "else:\n",
        "  #args.wDir = os.path.basename(target_file)\n",
        "  args.wDir = directory\n",
        "\n",
        "args.geneFile = target_file\n",
        "\n",
        "args.oligoLen = 250 #@param {type:\"integer\"}\n",
        "args.fragmentLen = \"auto\"  #@param {type:\"string\"}\n",
        "if args.fragmentLen == \"auto\":\n",
        "  args.fragmentLen = \"\"\n",
        "args.melting_temp_low = 58 #@param {type:\"integer\"}\n",
        "args.melting_temp_high = 62 #@param {type:\"integer\"}\n",
        "args.overlap = 4 #@param {type:\"integer\"}\n",
        "args.barcode_start = 0 #@param {type:\"integer\"}\n",
        "type_IIS_restriction_enzyme = \"BsmBI\" #@param [\"BsmBI\", \"BsaI\"]\n",
        "if type_IIS_restriction_enzyme == \"BsmBI\":\n",
        "  args.restriction_sequence = \"CGTCTC\"\n",
        "else:\n",
        "  args.restriction_sequence = \"GGTCTC\"\n",
        "sequences_to_avoid = \"CGTCTC,GGTCTC\" #@param {type:\"string\"}\n",
        "args.avoid_sequence = sequences_to_avoid.split(\", \")\n",
        "args.matchSequences = False\n",
        "\n",
        "#@markdown Codon Usage (If using custom usage you will see a button to upload file at the bottom)\n",
        "args.usage = \"human\" #@param [\"human\", \"ecoli\"]\n",
        "custom_codon_usage = False #@param {type:\"boolean\"}\n",
        "# input custom mutation file\n",
        "if custom_codon_usage:\n",
        "  print(\"Look in DIMPLE/data/custom_codon_usage.txt for example file\")\n",
        "  uploaded = files.upload()\n",
        "  usage_file = list(uploaded.keys())[0]\n",
        "  with open(usage_file) as f:\n",
        "    usage = f.readlines()\n",
        "  args.usage = ast.literal_eval(usage[0].strip('\\n'))\n",
        "  pd.DataFrame(args.usage,index=[0])\n",
        "\n",
        "#@markdown Select Mutations (Select one or more)\n",
        "select_domain_insertion = False #@param {type:\"boolean\"}\n",
        "if select_domain_insertion:\n",
        "  args.dis = True\n",
        "else:\n",
        "  args.dis = False\n",
        "args.handle = \"AGCGGGAGACCGGGGTCTCTGAGC\" #@param {type:\"string\"}\n",
        "select_insertions = False #@param {type:\"boolean\"}\n",
        "insertions = \"GGG,GGGGGG\" #@param {type:\"string\"}\n",
        "if select_insertions:\n",
        "  args.insertions = insertions.split(\",\")\n",
        "else:\n",
        "  args.insertions = \"\"\n",
        "select_deletions = False #@param {type:\"boolean\"}\n",
        "deletions = \"3,6\" #@param {type:\"string\"}\n",
        "if select_deletions:\n",
        "  args.deletions = deletions.split(\",\")\n",
        "else:\n",
        "  args.deletions = \"\"\n",
        "select_deep_mutation_scan = False #@param {type:\"boolean\"}\n",
        "if select_deep_mutation_scan:\n",
        "  args.dms = True\n",
        "else:\n",
        "  args.dms = False\n",
        "#@markdown Mutation Settings\n",
        "amino_acid_substitutions = \"Cys,Asp,Ser,Gln,Met,Asn,Pro,Lys,Thr,Phe,Ala,Gly,Ile,Leu,His,Arg,Trp,Val,Glu,Tyr\" #@param {type:\"string\"}\n",
        "args.include_stop_codons = False #@param {type:\"boolean\"}\n",
        "args.include_synonymous_mutations = False #@param {type:\"boolean\"}\n",
        "\n",
        "args.make_double_mutations = False #@param {type:\"boolean\"}\n",
        "select_custom_mutations = False #@param {type:\"boolean\"}\n",
        "if select_custom_mutations:\n",
        "  print(\"Look in DIMPLE/data/custom_codon_usage.txt for example file\")\n",
        "  uploaded = files.upload()\n",
        "  custom_mutations_file = list(uploaded.keys())[0]\n",
        "  with open(custom_mutations_file) as f:\n",
        "    lines = f.readlines()\n",
        "  args.custom_mutations = parse_custom_mutations(lines[1:])\n",
        "  pd.DataFrame(args.custom_mutations,index=[0])\n",
        "else:\n",
        "  args.custom_mutations = {}\n",
        "args.avoid_breaksites = False #@param {type:\"boolean\"}\n",
        "args.avoid_breaksites_list = \"\" #@param {type:\"string\"}\n",
        "\n",
        "#!python3 DIMPLE/run_dimple.py -geneFile=$target_file -oligoLen=$oligo_length -DMS=$select_DMS -usage=$codon_usage_bias -include_stop_codons=$include_stop_codons -restriction_sequence=$type_IIS_restriction_enzyme -avoid_sequence=$sequences_to_avoid -insertions=$insertions -deletions=$deletions -fragmentLen=$fragment_length -overlap=$fragment_overlap -barcode_start=$barcode_start_position -include_stop_codons=$include_stop_codons -usage=$custom_mutations\n",
        "\n",
        "\n",
        "# Check genetic handle for non nucleic characters\n",
        "if any([x not in ['A', 'C', 'G', 'T', 'a', 'c', 'g', 't'] for x in args.handle]):\n",
        "    raise ValueError('Genetic handle contains non nucleic bases')\n",
        "\n",
        "DIMPLE.handle = args.handle\n",
        "DIMPLE.synth_len = args.oligoLen\n",
        "if args.fragmentLen:\n",
        "    DIMPLE.maxfrag = args.fragmentLen\n",
        "else:\n",
        "    DIMPLE.maxfrag = args.oligoLen - 62 - args.overlap  # 62 allows for cutsites and barcodes\n",
        "\n",
        "DIMPLE.dms = args.dms\n",
        "\n",
        "#  adjust primer primerBuffer\n",
        "DIMPLE.primerBuffer += args.overlap\n",
        "\n",
        "DIMPLE.avoid_sequence = args.avoid_sequence\n",
        "DIMPLE.barcodeF = DIMPLE.barcodeF[int(args.barcode_start):]\n",
        "DIMPLE.barcodeR = DIMPLE.barcodeR[int(args.barcode_start):]\n",
        "DIMPLE.cutsite = Seq(args.restriction_sequence)\n",
        "DIMPLE.avoid_sequence = [Seq(x) for x in args.avoid_sequence]\n",
        "DIMPLE.stop_codon = args.include_stop_codons\n",
        "DIMPLE.make_double = args.make_double_mutations\n",
        "DIMPLE.usage = codon_usage(args.usage)\n",
        "\n",
        "OLS = addgene(os.path.join(args.wDir, args.geneFile).strip())\n",
        "if args.avoid_breaksites_list and select_custom_mutations:\n",
        "    OLS[0].problemsites = set(int(x) for x in args.custom_mutations.keys())\n",
        "    # add extras\n",
        "    if app.avoid_others_list.get() != '':\n",
        "        OLS[0].problemsites.update([int(x) for x in app.avoid_others_list.get().split(',')])\n",
        "    for i in range(len(OLS[0].breaksites)):\n",
        "        switch_fragmentsize(OLS[0], 1, OLS)\n",
        "if args.matchSequences == 'match':\n",
        "    align_genevariation(OLS)\n",
        "if args.deletions:\n",
        "    args.deletions = [int(x) for x in args.deletions]\n",
        "if not any([args.dis, args.dms, args.insertions, args.deletions]):\n",
        "    raise ValueError(\"Didn't select any mutations to generate\")\n",
        "print(args.dms)\n",
        "print(args.overlap)\n",
        "generate_DMS_fragments(OLS, args.overlap, args.overlap, args.include_synonymous_mutations, args.custom_mutations, args.dms, args.insertions, args.deletions, args.dis, args.wDir)\n",
        "\n",
        "post_qc(OLS)\n",
        "print_all(OLS, args.wDir)\n",
        "\n",
        "directory = args.wDir\n",
        "!zip -r file.zip $directory\n",
        "files.download(\"file.zip\")"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.10"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}