{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "view-in-github"
   },
   "source": [
    "<a href=\"https://colab.research.google.com/github/coywil26/DIMPLE/blob/master/DIMPLE.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "FHy8rovTWDVB"
   },
   "outputs": [],
   "source": [
    "#@title 1. Install\n",
    "\n",
    "#@markdown Please execute each cell by pressing the *Play* button on the left.\n",
    "\n",
    "# download github \"https://github.com/coywil26/DIMPLE\" and install dependencies\n",
    "!git clone https://github.com/coywil26/DIMPLE\n",
    "\n",
    "!pip install -r DIMPLE/requirements.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "dbA01B3wWfMP"
   },
   "outputs": [],
   "source": [
    "#@title 2. Upload target file\n",
    "#@markdown Please execute to upload file\n",
    "from google.colab import files\n",
    "\n",
    "uploaded = files.upload()\n",
    "\n",
    "for fn in uploaded.keys():\n",
    "  print('User uploaded file \"{name}\" with length {length} bytes'.format(\n",
    "      name=fn, length=len(uploaded[fn])))\n",
    "target_file = fn\n",
    "!mkdir workspace\n",
    "!mv $target_file workspace\n",
    "directory = \"workspace/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "J-ZYosDxWqZR"
   },
   "outputs": [],
   "source": [
    "#@title 3. Generate Mutation Fragments and Primers\n",
    "#@markdown Oligo Parameters\n",
    "from google.colab import files\n",
    "import os\n",
    "import argparse\n",
    "from DIMPLE.DIMPLE.DIMPLE import align_genevariation, print_all, post_qc, addgene, DIMPLE, generate_DMS_fragments\n",
    "from DIMPLE.DIMPLE.utilities import parse_custom_mutations\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "class arguments:\n",
    "  pass\n",
    "args = arguments()\n",
    "\n",
    "# if target_file doesnt exist use default test file\n",
    "if 'target_file' not in locals():\n",
    "  target_file = \"DIMPLE/tests/combined_fasta.fa\"\n",
    "  args.wDir = \"DIMPLE/tests/\"\n",
    "  print('test run')\n",
    "else:\n",
    "  #args.wDir = os.path.basename(target_file)\n",
    "  args.wDir = directory\n",
    "\n",
    "args.geneFile = target_file\n",
    "\n",
    "args.oligoLen = 250 #@param {type:\"integer\"}\n",
    "args.fragmentLen = \"auto\"  #@param {type:\"string\"}\n",
    "if args.fragmentLen == \"auto\":\n",
    "  args.fragmentLen = \"\"\n",
    "args.overlap = 4 #@param {type:\"integer\"}\n",
    "args.barcode_start = 0 #@param {type:\"integer\"}\n",
    "type_IIS_restriction_enzyme = \"BsmBI\" #@param [\"BsmBI\", \"BsaI\"]\n",
    "if type_IIS_restriction_enzyme == \"BsmBI\":\n",
    "  args.restriction_sequence = \"CGTCTC\"\n",
    "else:\n",
    "  args.restriction_sequence = \"GGTCTC\"\n",
    "sequences_to_avoid = \"CGTCTC, GGTCTC\" #@param {type:\"string\"}\n",
    "args.avoid_sequence = sequences_to_avoid.split(\", \")\n",
    "args.matchSequences = False\n",
    "\n",
    "#@markdown Codon Usage (If using custom usage you will see a button to upload file at the bottom)\n",
    "args.usage = \"human\" #@param [\"human\", \"ecoli\"]\n",
    "custom_codon_usage = False #@param {type:\"boolean\"}\n",
    "# input custom mutation file\n",
    "if custom_codon_usage:\n",
    "  usage_file = files.upload()\n",
    "  with open(usage_file) as f:\n",
    "    usage = f.readlines()\n",
    "  args.usage = ast.literal_eval(usage.strip('\\n'))\n",
    "\n",
    "#@markdown Select Mutations\n",
    "select_domain_insertion = False #@param {type:\"boolean\"}\n",
    "if select_domain_insertion:\n",
    "  args.dis = True\n",
    "else:\n",
    "  args.dis = False\n",
    "args.handle = \"AGCGGGAGACCGGGGTCTCTGAGC\" #@param {type:\"string\"}\n",
    "select_insertions = False #@param {type:\"boolean\"}\n",
    "insertions = \"GGG,GGGGGG\" #@param {type:\"string\"}\n",
    "if select_insertions:\n",
    "  args.insertions = insertions.split(\",\")\n",
    "else:\n",
    "  args.insertions = \"\"\n",
    "select_deletions = False #@param {type:\"boolean\"}\n",
    "deletions = \"3,6\" #@param {type:\"string\"}\n",
    "if select_deletions:\n",
    "  args.deletions = deletions.split(\",\")\n",
    "else:\n",
    "  args.deletions = \"\"\n",
    "select_deep_mutation_scan = False #@param {type:\"boolean\"}\n",
    "if select_deep_mutation_scan:\n",
    "  args.dms = True\n",
    "else:\n",
    "  args.dms = False\n",
    "amino_acid_substitutions = \"Cys,Asp,Ser,Gln,Met,Asn,Pro,Lys,Thr,Phe,Ala,Gly,Ile,Leu,His,Arg,Trp,Val,Glu,Tyr\" #@param {type:\"string\"}\n",
    "args.include_stop_codons = False #@param {type:\"boolean\"}\n",
    "args.include_synonymous_mutations = False #@param {type:\"boolean\"}\n",
    "select_custom_mutations = False #@param {type:\"boolean\"}\n",
    "if select_custom_mutations:\n",
    "  custom_mutations_file = files.upload()\n",
    "  with open(custom_mutations_file) as f:\n",
    "    lines = f.readlines()\n",
    "  args.custom_mutations = parse_custom_mutations(lines)\n",
    "else:\n",
    "  args.custom_mutations = {}\n",
    "args.make_double_mutations = False #@param {type:\"boolean\"}\n",
    "\n",
    "#!python3 DIMPLE/run_dimple.py -geneFile=$target_file -oligoLen=$oligo_length -DMS=$select_DMS -usage=$codon_usage_bias -include_stop_codons=$include_stop_codons -restriction_sequence=$type_IIS_restriction_enzyme -avoid_sequence=$sequences_to_avoid -insertions=$insertions -deletions=$deletions -fragmentLen=$fragment_length -overlap=$fragment_overlap -barcode_start=$barcode_start_position -include_stop_codons=$include_stop_codons -usage=$custom_mutations\n",
    "\n",
    "\n",
    "# Check genetic handle for non nucleic characters\n",
    "if any([x not in ['A', 'C', 'G', 'T', 'a', 'c', 'g', 't'] for x in args.handle]):\n",
    "    raise ValueError('Genetic handle contains non nucleic bases')\n",
    "\n",
    "#DIMPLE.handle = args.handle\n",
    "DIMPLE.synth_len = args.oligoLen\n",
    "if args.fragmentLen:\n",
    "    DIMPLE.maxfrag = args.fragmentLen\n",
    "else:\n",
    "    DIMPLE.maxfrag = args.oligoLen - 62 - args.overlap  # 62 allows for cutsites and barcodes\n",
    "\n",
    "DIMPLE.dms = args.dms\n",
    "\n",
    "#  adjust primer primerBuffer\n",
    "DIMPLE.primerBuffer += args.overlap\n",
    "\n",
    "DIMPLE.avoid_sequence = args.avoid_sequence\n",
    "DIMPLE.barcodeF = DIMPLE.barcodeF[int(args.barcode_start):]\n",
    "DIMPLE.barcodeR = DIMPLE.barcodeR[int(args.barcode_start):]\n",
    "DIMPLE.cutsite = Seq(args.restriction_sequence)\n",
    "DIMPLE.avoid_sequence = [Seq(x) for x in args.avoid_sequence]\n",
    "DIMPLE.stop_codon = args.include_stop_codons\n",
    "DIMPLE.make_double = args.make_double_mutations\n",
    "if args.usage == 'ecoli':\n",
    "    DIMPLE.usage = {\n",
    "        'TTT': 0.58, 'TTC': 0.42, 'TTA': 0.14, 'TTG': 0.13, 'TAT': 0.59, 'TAC': 0.41, 'TAA': 0.61, 'TAG': 0.09,\n",
    "        'CTT': 0.12, 'CTC': 0.1, 'CTA': 0.04, 'CTG': 0.47, 'CAT': 0.57, 'CAC': 0.43, 'CAA': 0.34, 'CAG': 0.66,\n",
    "        'ATT': 0.49, 'ATC': 0.39, 'ATA': 0.11, 'ATG': 1, 'AAT': 0.49, 'AAC': 0.51, 'AAA': 0.74, 'AAG': 0.26,\n",
    "        'GTT': 0.28, 'GTC': 0.2, 'GTA': 0.17, 'GTG': 0.35, 'GAT': 0.63, 'GAC': 0.37, 'GAA': 0.68, 'GAG': 0.32,\n",
    "        'TCT': 0.17, 'TCC': 0.15, 'TCA': 0.14, 'TCG': 0.14, 'TGT': 0.46, 'TGC': 0.54, 'TGA': 0.3, 'TGG': 1,\n",
    "        'CCT': 0.18, 'CCC': 0.13, 'CCA': 0.2, 'CCG': 0.49, 'CGT': 0.36, 'CGC': 0.36, 'CGA': 0.07, 'CGG': 0.11,\n",
    "        'ACT': 0.19, 'ACC': 0.4, 'ACA': 0.17, 'ACG': 0.25, 'AGT': 0.16, 'AGC': 0.25, 'AGA': 0.07, 'AGG': 0.04,\n",
    "        'GCT': 0.18, 'GCC': 0.26, 'GCA': 0.23, 'GCG': 0.33, 'GGT': 0.35, 'GGC': 0.37, 'GGA': 0.13, 'GGG': 0.15\n",
    "    }  # E.coli codon usage table\n",
    "elif args.usage == 'human':\n",
    "    DIMPLE.usage = {\n",
    "        'TTT': 0.45, 'TTC': 0.55, 'TTA': 0.07, 'TTG': 0.13, 'TAT': 0.43, 'TAC': 0.57, 'TAA': 0.28, 'TAG': 0.2,\n",
    "        'CTT': 0.13, 'CTC': 0.2, 'CTA': 0.07, 'CTG': 0.41, 'CAT': 0.41, 'CAC': 0.59, 'CAA': 0.25, 'CAG': 0.75,\n",
    "        'ATT': 0.36, 'ATC': 0.48, 'ATA': 0.16, 'ATG': 1, 'AAT': 0.46, 'AAC': 0.54, 'AAA': 0.42, 'AAG': 0.58,\n",
    "        'GTT': 0.18, 'GTC': 0.24, 'GTA': 0.11, 'GTG': 0.47, 'GAT': 0.46, 'GAC': 0.54, 'GAA': 0.42, 'GAG': 0.58,\n",
    "        'TCT': 0.18, 'TCC': 0.22, 'TCA': 0.15, 'TCG': 0.06, 'TGT': 0.45, 'TGC': 0.55, 'TGA': 0.52, 'TGG': 1,\n",
    "        'CCT': 0.28, 'CCC': 0.33, 'CCA': 0.27, 'CCG': 0.11, 'CGT': 0.08, 'CGC': 0.19, 'CGA': 0.11, 'CGG': 0.21,\n",
    "        'ACT': 0.24, 'ACC': 0.36, 'ACA': 0.28, 'ACG': 0.12, 'AGT': 0.15, 'AGC': 0.24, 'AGA': 0.2, 'AGG': 0.2,\n",
    "        'GCT': 0.26, 'GCC': 0.4, 'GCA': 0.23, 'GCG': 0.11, 'GGT': 0.16, 'GGC': 0.34, 'GGA': 0.25, 'GGG': 0.25\n",
    "    }\n",
    "else:\n",
    "    DIMPLE.usage = args.usage\n",
    "\n",
    "OLS = addgene(os.path.join(args.wDir, args.geneFile).strip())\n",
    "\n",
    "if args.matchSequences == 'match':\n",
    "    align_genevariation(OLS)\n",
    "if args.deletions:\n",
    "    args.deletions = [int(x) for x in args.deletions]\n",
    "if not any([args.dms, args.insertions, args.deletions]):\n",
    "    raise ValueError(\"Didn't select any mutations to generate\")\n",
    "print(args.dms)\n",
    "print(args.overlap)\n",
    "generate_DMS_fragments(OLS, args.overlap, args.overlap, args.include_synonymous_mutations, args.custom_mutations, args.dms, args.insertions, args.deletions, args.dis, args.wDir)\n",
    "\n",
    "post_qc(OLS)\n",
    "print_all(OLS, args.wDir)\n",
    "\n",
    "directory = args.wDir\n",
    "!zip -r file.zip $directory\n",
    "files.download(\"file.zip\")"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "authorship_tag": "ABX9TyMMNXDUaQ6ReB+IW1kJN6kT",
   "include_colab_link": true,
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
